implementation/servlet/src/main/java/net/opentsdb/servlet/filter/EnvoyHeaderAuthFilter.java

```diff
@@ -0,0 +1,207 @@
+import org.slf4j.LoggerFactory;
+
+
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletRequestWrapper;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.security.Principal;
+
+/**
+ * A filter that processes MTLS certificates from Athenz to authenticate a
+ * caller.
+ *
+ * @since 3.0
+ */
+public class EnvoyHeaderAuthFilter extends BaseAuthenticationPlugin {
+  protected static final Logger LOG = LoggerFactory.getLogger(
+          EnvoyHeaderAuthFilter.class);
+  public static final String ENVOY_HEADER_KEY = "envoy_header_key";
+  public static final String TYPE = "EnvoyHeaderAuthFilter";
+  public static final String TOKEN_TYPE = "X509";
+  public Principal principal;
+
+
+  @Override
+  public Deferred<Object> initialize(final TSDB tsdb, final String id) {
+    return super.initialize(tsdb, id).addCallback(new Callback<Object, Object>() {
+      @Override
+      public Object call(Object arg) throws Exception {
+        registerConfigs(tsdb);
+        LOG.info("Successfully loaded Envoy Header authentication filter.");
+        return null;
+      }
+    });
+  }
+
+  @Override
+  protected void registerConfigs(final TSDB tsdb) {
+    super.registerConfigs(tsdb);
+    Configuration config = tsdb.getConfig();
+    if (!config.hasProperty(ENVOY_HEADER_KEY)) {
+      config.register(ENVOY_HEADER_KEY, "x-forwarded-client-cert", true,
+              "The header name of envoy certificates.");
+    }
+
+  }
+
+  @Override
+  public Authorization authorization() {
+    return null;
+  }
+
+  @Override
+  public String type() {
+    return TYPE;
+  }
+
+  @Override
+  protected void runFilter(final ServletRequest servletRequest,
+                           final ServletResponse servletResponse,
+                           final FilterChain chain) throws IOException, ServletException {
+
+    final AuthState state = authenticate(servletRequest);
+
+    if (state != null) {
+      servletRequest.setAttribute(AUTH_STATE_KEY, state);
+      HttpServletRequestWrapper wrapper =
+              new HttpServletRequestWrapper((HttpServletRequest) servletRequest) {
+                @Override
+                public java.security.Principal getUserPrincipal() {
+                  return state.getPrincipal();
+                }
+              };
+      chain.doFilter(wrapper, servletResponse);
+    }
+    else {
+      sendResponse((HttpServletResponse) servletResponse, 403,
+              "Missing or invalid header ");
+    }
+  }
+
+  @Override
+  public AuthState authenticate(final ServletRequest servletRequest) {
+
+    String header=((HttpServletRequest ) servletRequest ).getHeader(tsdb.getConfig().getString(ENVOY_HEADER_KEY));
+    if (header != null){
+      String commonName=header.split(";")[2].split(",")[0].split("=")[2];
```

Comment:
This can throw NPEs if the header is malformed. Split it into multiple steps wherein you split, validate the result is the proper size, then pick the field you want, validate it's name, then split if applicable, repeat till you get the value you want. If a field is missing or invalid then you can log at the debug level and return a null.

Reply:
Yes, ack.!